extensions/blocks/repeat-visitor/components/edit.js

```diff
@@ -34,7 +34,7 @@ class RepeatVisitorEdit extends Component {
 	setCriteria = criteria => this.props.setAttributes( { criteria } );
 	setThreshold = threshold => {
 		if ( /^\d+$/.test( threshold ) && +threshold > 0 ) {
-			this.props.setAttributes( { threshold, isThresholdValid: true } );
+			this.props.setAttributes( { threshold: +threshold, isThresholdValid: true } );
```

Comment:
While we're at it -- can we get rid of the `isThresholdValid` attribute? :grimacing: I find it kinda weird to store an attribute that contains information about the validity of another (`threshold`) -- shouldn't we be able to infer validity of `threshold` from its value at all times?

Reply:
Thanks for the suggestion @ockham - makes sense. I'd prefer that we address it in a followup PR after the release if that's alright.

Reply:
The thing is, if we release the block _with_ this attribute and retire it afterwards, we'll need to write deprecations for it, since removing then it will otherwise invalidate our user's existing blocks at that point. If we manage to remove it before the release, we won't have to do that :slightly_smiling_face:

Reply:
That's a fair point @ockham - I hadn't considered needing to write a deprecation.

I've had a tinker with the code, and I think we do need to store somewhere that we're currently in an error state. If we get a new threshold set, if it's invalid (like '-5') we don't want to set it in `this.props.attributes.threshold` because then it could be preserved. That's why we've been using a separate attribute (`isThresholdValid`) to keep track of whether the last-entered value was acceptable or not, without storing it.

Would a rename of the attribute help? Is `isError` clearer?

cc @jsnmoon

Reply:
@bluefuton note that these attributes can be modified directly in the post content, so checking for validity (or lack of the attribute) each time the value is read is a good idea anyway.

Reply:
> Would a rename of the attribute help? Is isError clearer?

The concern is that it seems out of place to use attributes for this. Attributes are serialized and stored, so we're storing invalid data, as well as (meta)data that the value is invalid.

I'd suggest that we shouldn't serialize an invalid attribute, and we should be able to compute validity on the fly. That seems to lead to juggling to levels of state: block attributes and local component state.

- The attribute stores the valid value.
- Changes should be pushed to the attribute (setAttribute) there as soon as they're validated. "Dirty" component state (invalid or missing values) should be tracked in local component state. This may mean clearing the attribute when the form is dirty.

In time, it would be nice to land on a solution we can share across components. In Simple Payments, we were exploring Formik to track form state, independent of attributes, but were unable to finish those changes due to time constraints.

Reply:
I was able to replace `this.props.attributes.isThresholdValid` with `this.state.isThresholdValid`. 

It seems to work as expected - let me know what you all think!

Reply:
Thanks a lot @jsnmoon, glad to hear you could replace the attribute with component state! :tada: 

Out of curiosity, with these properties set https://github.com/Automattic/jetpack/blob/cda78ee745f0d6cf8dd85176aa10b4ccbbe18e2b/extensions/blocks/repeat-visitor/components/edit.js#L84-L87 -- how can I actually repro entering an invalid value into that field? :sweat_smile:

Reply:
@ockham Testers in our call for testing figured out multiple ways :)

You can try inputs like `1+2`, `8eee`, `-14`, and so on.

Reply:
Ah, I see, thanks :slightly_smiling_face: 

Let's remove the `pattern` attr then since that's apparently [not supported](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number#Pattern_validation) for this type of input field.